### ISSUE TITLE
[js_interop] assertInterop is not injected by DDC for JSInterop calls within dart:* libraries

### ISSUE THREAD

**Issue by joshualitt** [dart-lang/sdk#48908]
This is just a tracking bug for a strange issue I noticed while porting Flutter Web to static interop.

They have a class, `JsUrlStrategy` that acts as a bridge between the App and the Flutter Web engine. Dart runs on both sides of this bridge. For some reason, a Dart function escapes through the interop boundary without having been wrapped in `allowInterop`. I only caught this issue because I was migrating the class to static interop, where the assert fires.

This behavior might make sense in Dart2js, where the assert only fires when the callback is invoked in JS which does not happen in this case. However, it doesn't really make sense for DDC.

Because we are migrating this codebase to static interop, and because the issue does not exist there, I did not investigate further at this point, but we may want to investigate down the road if we see this issue become more pervasive during migration.

**Comment by sigmundch** [comment]
For clarity. This issue revealed a strange inconsistency, which is why we are trying to track it down. That is, DDC had one behavior with the traditional JSInterop API, but saw failures when using static JSInterop, and we didn't expect this.

This caused a change to be reverted and the fix was to add an `allowInterop` where there wasn't one before. 

Pointers:
* initial PR https://github.com/flutter/engine/pull/32683
* revert https://github.com/flutter/engine/pull/32919
* Failure logs https://ci.chromium.org/ui/p/flutter/builders/try/Linux%20web_long_running_tests_3_5/15692/overview
* reland https://github.com/flutter/engine/pull/32928

**Comment by sigmundch** [comment]
I have convinced myself now that this issue is pretty contained. I was still worried that https://github.com/flutter/engine/pull/32683, which only changes to use `@staticInterop`,  could mean that other users could break in similar ways, but upon closer inspection, I believe this only affects the Flutter web SDK.

The reason is that DDC inserts `assertInterop` automatically when calling interop methods and passing arguments whose type could be a function, but it has one accidental exception: if the target lives in the SDK, we don't. This is because earlier on we gave special meaning to `external` in the SDK. See https://github.com/dart-lang/sdk/blob/72c153b93084ead3e9d2079b2f3c9568d8fa2d09/pkg/dev_compiler/lib/src/kernel/js_interop.dart#L43 which is queried before inserting the `assertInterop` calls in the compiler. For example in https://github.com/dart-lang/sdk/blob/72c153b93084ead3e9d2079b2f3c9568d8fa2d09/pkg/dev_compiler/lib/src/kernel/compiler.dart#L5643

The static JS interop implementation just translates code into calls to js_util directly, which by default always has this check and will not omit the check even if the definition was found in a `dart:*` library.

So, because the code that had the discrepancy lives in `dart:ui`, we ended up hitting this corner case.

As for action items: we could prevent similar issues in the future by updating the definition of `isJsMember` to be more robust. That said, I'm OK ignoring this issue now that we know that its effect is confined to only the flutter SDK and wont affect the broader set of users of JS interop.

@joshualitt @srujzs - if you agree, I'm inclined to close this and leave it as is.

**Comment by srujzs** [comment]
I think I'm fine leaving the discrepancy as is for now (and maybe reword the title so it's more discoverable if we ever come across this again).

This likely won't pop up again after we migrate Flutter web (since we don't use non-`@staticInterop` `package:js` as much anywhere else in the `dart:*` libraries), and most likely we won't be inclined to fix this as we push forward with the new interop.
